### PR TITLE
ActiveRecord query optimizations

### PR DESCRIPTION
**Pull request by wjordan**
This PR contains a number of ActiveRecord query optimizations for several high-scale routes.

Using the db-query integration tests being added in #18403, query-counts on our high-scale routes have been reduced as follows:

- `milestone`: **11** => **6**
- `user_progress`: **7** => **4**
- `script_levels#show`: **22** => **9**

Each individual commit contains a separate individual optimization, with a separate adjustment to the query-counts.

(Note: this PR depends on #18403 and is based on top of that PR's commits - until that PR is merged, see [`db-query-test...db-fixes`](https://github.com/code-dot-org/code-dot-org/compare/db-query-test...db-fixes) for a diff of just the new commits.)

### CODE REVIEW

File: dashboard/config/initializers/atomic_save.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Provides `#atomic_save!` which skips the default transaction
+# wrapping behavior of `#save!`.
+#
+# Only use in cases where validations/callbacks will never
+# need to trigger any rollback of the operation.
+module ActiveRecord
+  module Persistence
+    alias_method :atomic_save!, :save!
+  end
+end
```
Comment (wjordan) [inline dashboard/config/initializers/atomic_save.rb:10]: See also #15282 for some previous [discussion](https://github.com/code-dot-org/code-dot-org/pull/15282/files#r117862179) on this addition.

**Comment by Bjvanminnen** [comment]
Is it worth splitting these optimizations into a couple different PRs so that if something goes wrong we can better determine root cause?

**Comment by wjordan** [comment]
> Is it worth splitting these optimizations into a couple different PRs so that if something goes wrong we can better determine root cause?

Since tests are passing, splitting these individual commits into different PRs wouldn't be useful unless they're actually deployed/monitored separately in production. Staging all 8 of these optimizations across different production deployments in order to deploy/monitor them separately for something going wrong is a _lot_ of extra effort I don't think is worth it. None of these optimizations ~are terribly complex~ change too much code, and each has already been verified with the db-query integration-test coverage I've added. I don't think it will be difficult to determine the root cause of something going wrong with any of them even when deployed together.

**Comment by wjordan** [comment]
After more consideration I'm going to split up this PR for code-reviewing each optimization separately. I won't ensure each PR gets deployed/monitored separately though, so they may still be deployed together.

(I need to close/reopen this PR anyway, because it doesn't seem to be picking up any changes to the branch since #18403 was merged for some reason.. [edit: this might be related to a GitHub-service incident.])

**Pull request closed without merging**